http/README.md

```diff
@@ -0,0 +1,133 @@
+[@thomaseizinger]: https://github.com/thomaseizinger
+
+## Introduction
+
+This document defines how libp2p nodes can offer and use an HTTP transport alongside their other transports to support application protocols with HTTP semantics. This allows a wider variety of nodes to participate in the libp2p network, for example:
+
+- Browsers communicating with other libp2p nodes without needing a WebSocket, WebTransport, or WebRTC connection.
+- HTTP only edge workers can run application protocols and respond to peers on the network.
+- `curl` from the command line can make requests to other libp2p nodes.
+
+The HTTP transport will also allow application protocols to make use of HTTP intermediaries such as HTTP caching, and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:
+
+- Connectivity – Work on top of WebRTC, WebTransport, QUIC, TCP, or an HTTP transport.
+- Hole punching – Work with peers behind NATs.
+- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
+- Peer discovery – Learn about a peer given their peer id.
+
+## HTTP Semantics vs Encodings vs Transport
+
+HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.
+
+
+```mermaid
+graph TB
+    subgraph "HTTP Semantics"
+        HTTP
+    end
+    subgraph "Encoding"
+        HTTP1.1[HTTP/1.1]
+        HTTP2[HTTP/2]
+        HTTP3[HTTP/3]
+    end
+    subgraph "Transports"
+        Libp2p[libp2p streams]
+        HTTPTransport[HTTP transport]
+    end
+    HTTP --- HTTP1.1
+    HTTP --- HTTP1.1
+    HTTP1.1 --- Libp2p
+    HTTP --- HTTP2
+    HTTP --- HTTP3
+    HTTP1.1 --- HTTPTransport
+    HTTP2 --- HTTPTransport
+    HTTP3 --- HTTPTransport
+```
+
+- *HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is
+  the stateless application-level protocol that you work with when writing HTTP
+  apis (for example).
+
+- *HTTP encoding* is the thing that takes your high level request/response
+  defined in terms of HTTP semantics and encodes it into a form that can be sent
+  over the wire.
+
+- *HTTP transport* is the thing that takes your encoded reqeust/response and
+  sends it over the wire. For HTTP/1.1 and HTTP/2, this is a TCP+TLS connection.
+  For HTTP/3, this is a QUIC connection.
+
+When this document says *HTTP* it is generally referring to *HTTP semantics*.
+
+## Interoperability with existing HTTP systems
+
+A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in
+their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
```

Comment:
Peer ID authentication is optional, but if you have to add a PeerID in the sample multiaddr where does it fit?

There has been a lot of discussion about HTTP multiaddrs being path multiaddrs in the past, which would prevent `/p2p/Qmfoo` being added onto the end, though I don't think they were ever resolved.

Reply:
I have an open issue for how to document HTTP resources: https://github.com/libp2p/specs/issues/554 and a corresponding PR: https://github.com/multiformats/multiaddr/pull/164

With that change it would be very much like it is today:
`/dns4/example.com/tls/http/p2p/Qmfoo` or `/dns4/example.com/tls/http/httppath/somepath/p2p/Qmfoo`

This is consistent with interpreting multiaddrs [right to left](https://github.com/multiformats/multiaddr/?tab=readme-ov-file#interpreting-multiaddrs), as we first see the peer we are communicating with, the (optional) HTTP resource we are referencing, and the direction to connect to the peer.

Reply:
Makes sense 👍 

https://github.com/multiformats/multiaddr/pull/164 suggests adding the `httppath` component after the Peer ID, which seems to contradict the above.  Maybe it needs updating?

> `/httppath` should be appended to the end of an existing multiaddr, including after the peer id component (p2p). As an example, here's a multiaddr referencing the `.well-known/libp2p` HTTP resource along with a way to reach that peer:
>
> ```
> /ip4/1.2.3.4/tcp/443/tls/http/p2p/12D.../httppath/.well-known%2Flibp2p
> ```

Source: https://github.com/multiformats/multiaddr/pull/164/files#diff-940635fb4ac11927020564dd7f96ac0604da98bb11aee1273030de75a31dcf06R27

Reply:
> https://github.com/multiformats/multiaddr/pull/164 suggests adding the httppath component after the Peer ID, which seems to contradict the above. Maybe it needs updating?

Good catch. Maybe my past self was right and my current self is wrong. I know my past self spent a lot more time thinking about that document than my current self did in writing that comment.

Reply:
My past self was right. and my comment [above](https://github.com/libp2p/specs/pull/508#discussion_r1525335920) is wrong.

Multiaddrs are interpreted right to left, but the peer-id is more about *where* this resource is located. We first learn about the HTTP resource we're interested in, and then learn about how to get it.